test/python/quantum_info/operators/symplectic/test_pauli_list.py

```diff
@@ -1999,6 +1999,32 @@ def test_evolve_clifford_qargs(self, phase):
         ]
         self.assertListEqual(value, target)
 
+    def test_group_qubit_wise_commuting(self):
+        """Test grouping qubit-wise commuting operators"""
+        input_labels = ["IY", "ZX", "XZ", "YI", "YX", "YY", "YZ", "ZI", "ZX", "ZY", "iZZ", "II"]
+        np.random.shuffle(input_labels)
+        pauli_list = PauliList(input_labels)
+        groups = pauli_list.group_qubit_wise_commuting()
+
+        # checking that every input Pauli in pauli_list is in a group in the ouput
+        assert all(((pauli in group) for group in groups) for pauli in pauli_list)
+
+        # checking that for every pair of groups in the output, there is at least one element of
+        # one group which does not commute with at least one element of the other group
+        mat1 = [
+            np.array(
+                [op.z + 2 * op.x for op in group],
+                dtype=np.int8,
+            )
+            for group in groups
+        ]
+        mat2 = [mat[:, None] for mat in mat1]
+
+        # value[i][j] will be False if groups i and j have non-commuting elements, True otherwise
+        value = np.array([[(i * j * (i - j) == 0).all(axis=(2, 1, 0)) for i in mat1] for j in mat2])
+        target = np.diag([True] * len(mat1))
+        assert (value == target).all()
```

Comment:
I tested your changes, but they caused an assertion error for groups that was produced like this: 
`groups = [PauliList(['II', 'XZ']),
 PauliList(['YZ']),
 PauliList(['ZX', 'ZX']),
 PauliList(['IY', 'ZY']),
 PauliList(['iZZ', 'ZI']),
 PauliList(['YX', 'YI']),
 PauliList(['YY'])] `
I think this is because the `commutes()` function that is used checks for general commutativity, not Qubit-wise commutativity. This PR was factoring out the grouper function out of `AbelianGrouper` which uses the Qubit-wise commuting definition. So, from what I understand, general commutativity would say that that the group `PauliList(['II', 'XZ’])` and ` PauliList(['ZX', 'ZX’])` does not have elements in the first group that do not commute with the second group, but the Qubit-wise commuting definition says that there are elements in the first group that do not Qubit-wise commute with elements in the second group (`Pauli(“XZ”)` does not Qubit-wise commute with `Pauli(“ZX”)`). I could have misinterpreted something, so please let me know if I have anything wrong! I see what you are saying that the tests should obviously check correct behavior, so I can try to rewrite the test to work with Qubit-wise commuting but also be more readable, or like you suggested, we could test with an example that has only one partition.

Reply:
Ah, sorry, this is my fault for missing that `Pauli.commutes` isn't the same type of commutation.  Good catch, too! In this case, would having a helper function
```python
def qubitwise_commutes(left: Pauli, right: Pauli):
    return len(left) == len(right) and all(a.commutes(b) for a, b in zip(left, right))
```
replace all the `pauli1.commutes(pauli2)` with `qubitwise_commutes(pauli1, pauli2)` work?

Reply:
That is an option I was thinking about earlier, but because the numpy implementation of the actual `group_qubit_wise_commuting()` doesn't use the function, I wasn't sure about introducing a `qubit_wise_commutes()` function in the test. But this definitely works too. I was also wondering if this would work (this is just a quick edit to your previous suggestion changes that would instead check for qubit_wise):
```python 
for group1, group2 in itertools.combinations(groups, 2):
    assert not all(
        all((single_pauli_1.commutes(single_pauli_2)) 
            for single_pauli_1,single_pauli_2 in itertools.product(group1_pauli,group2_pauli)
           )
        for group1_pauli, group2_pauli in itertools.product(group1, group2)
    )
```
If this doesn't work or isn't readable, I can just add the helper function you suggested.

Reply:
It's not a bad thing for the test to contain a different implementation - if anything, it's bad if the test uses _exactly_ the same method, because then there's circular logic.  The test method should focus on being very very readable, even at the cost of being much less efficient, as is the case here.  The numpy method will be much much faster, so that's the right thing to have in the production code, but the tests are only going to run on very small instances where it doesn't matter that the test scaling is poor.

Personally I'd prefer to wrap it up in a helper function (you can even define the helper inside the test itself):
```python
def test_group_paulis(self):
    def qubitwise_commutes(...):
        return ...
    
    input_list = [...]
    ....
```
because then the actual assert loop is much easier to read (it's hard to read nested `all` statements).

Reply:
That makes sense and it's good to know, thank you! I'll add that in.